kratos/includes/properties.h

```diff
@@ -278,31 +293,303 @@ class Properties : public IndexedObject
+
+    /**
+     * @brief This method gets the subproperty from the index corresponding to the property id (constant version)
+     * @param SubPropertyIndex The index of the subproperty to be get
+     * @return The pointer to the subproperty of interest
+     */
+    Properties::Pointer pGetSubPropertiesById(const IndexType SubPropertyIndex) const
+    {
+        // Looking into the database
+        const auto property_iterator = mSubPropertiesList.find(SubPropertyIndex);
+        if (property_iterator != mSubPropertiesList.end()) {
+            return *(property_iterator.base());
+        } else {
+            KRATOS_ERROR << "Subproperty ID: " << SubPropertyIndex << " is not defined on the current Properties ID: " << this->Id() << " creating a new one with ID: " << SubPropertyIndex << std::endl;
+            return nullptr;
+        }
+    }
+
+    /**
+     * @brief This method gets the subproperty from the index corresponding to the property id
+     * @param SubPropertyIndex The index of the subproperty to be get
+     * @return The reference to the subproperty of interest
+     */
+    Properties& GetSubPropertiesById(const IndexType SubPropertyIndex)
+    {
+        // Looking into the database
+        auto property_iterator = mSubPropertiesList.find(SubPropertyIndex);
+        if (property_iterator != mSubPropertiesList.end()) {
+            return *(property_iterator);
+        } else {
+            KRATOS_ERROR << "Subproperty ID: " << SubPropertyIndex << " is not defined on the current Properties ID: " << this->Id() << " creating a new one with ID: " << SubPropertyIndex << std::endl;
+            return *this;
+        }
+    }
+
+    /**
+     * @brief This method gets the subproperty from the index corresponding to the property id (constant version)
+     * @param SubPropertyIndex The index of the subproperty to be get
+     * @return The reference to the subproperty of interest
+     */
+    Properties& GetSubPropertiesById(const IndexType SubPropertyIndex) const
+    {
+        // Looking into the database
+        if (mSubPropertiesList.find(SubPropertyIndex) != mSubPropertiesList.end()) {
+            return *(mSubPropertiesList.find(SubPropertyIndex));
+        } else {
+            KRATOS_ERROR << "Subproperty ID: " << SubPropertyIndex << " is not defined on the current Properties ID: " << this->Id() << std::endl;
+        }
+    }
+
+    /**
+     * @brief This method checks if the subproperty exists from the adress corresponding to the proper structure of indexes
+     * @param rAdress The text that indicates the structure of subproperties to iterate and found the property of interest
+     * @return True if there is such subproperty, false otherwise
+     */
+    bool HasSubPropertiesByAddress(const std::string& rAdress)
+    {
+        const auto indexes = TrimComponentName(rAdress);
+        if (indexes[0] == this->Id()) {
+            Properties::Pointer p_property_interest = Kratos::make_shared<Properties>(*this);
```

Comment:
Why you need this pointer?

Reply:
In order to do the loop without additional changes, if the index list is only one index it returns the current property.

Reply:
Now with the last fix makes sense

Reply:
But the problem is that going out from this method you are deleting this as the only shared_ptr to it is out of scope. Note that Properties is not using intrusive pointer.

Reply:
Yes, I certainly don't know hot to solve this.